React Pages Migration: Account Settings

purfectliterature opened the pull request:
Part of #5098.

| Before | After |
| - | - |
| ![image](https://user-images.githubusercontent.com/51525686/197328946-77dc950d-122d-410e-ab52-a2ebd6455f4e.png) | ![image](https://user-images.githubusercontent.com/51525686/197328919-62f366a3-edf4-4274-9152-c9db27d69e54.png) |
| ![image](https://user-images.githubusercontent.com/51525686/197329015-b54c0b42-3870-40fb-9065-57405067b24a.png) | ![image](https://user-images.githubusercontent.com/51525686/197329096-6bf0a5e1-7a24-4b6c-b2ba-11f754731baa.png) |
| ![image](https://user-images.githubusercontent.com/51525686/197328909-225d0b6f-96e7-4548-bade-79668192a48c.png) | ![image](https://user-images.githubusercontent.com/51525686/197328963-d11201e5-2c28-46d1-9e8d-ed47d21a3bc3.png)

## Notes
### The second parameter of yup's `object().shape`
The second parameter `exclude?: [string, string][]` (as of `yup/lib/object.d.ts`) or `noSortEdges?: Array<[string, string]>` (as of [yup's method signature](https://github.com/jquense/yup/tree/pre-v1#objectshapefields-object-nosortedges-arraystring-string-schema)) allows us to specify keys that may have cyclic dependencies. As far as I know, it was not documented in [yup's documentation](https://github.com/jquense/yup/tree/pre-v1). It was mentioned in this discussion https://github.com/jquense/yup/issues/79#issuecomment-699605408, and I am putting it here for future reference.

### `dirtyFields` and `isDirty` behaviour of react-hook-form
- `isDirty` is determined by deep equal of the form's current values against the `defaultValues`.
- `dirtyFields` is determined at field level [due to performance considerations](https://github.com/react-hook-form/react-hook-form/issues/7970#issuecomment-1059724053).

As such, I have decided to NOT use `dirtyFields` because I found it to be unreliable. Our `Form` still uses `isDirty` to determine the dirty state and when to pop the heads-up up.

>**Note** The new `submitsDirtyFieldsOnly` prop works by comparing the form's current values against the `defaultValues`, similar to `isDirty`, **except without deep equal, but with strict equality, i.e., `===`**, due to performance concerns. This should generally be a non-issue since our fields are usually primitives anyway, but if we ever need to handle this, we could look into using [fast-deep-equal](https://github.com/epoberezkin/fast-deep-equal). Notably, objects and arrays are always going to be submitted `onSubmit` regardless of modifications. I guess we could add an `excludes?` parameter to ignore these fields, but I have yet to find a use case for this.

### Understanding the new `Form` methods
This PR extends `Form` with 4 new features:
- `submitsDirtyFieldsOnly?` prop, to support forms with multiple endpoints and big forms to only send necessary deltas for PATCHing.
- `resetByMerging: (data) => void`, to support forms with multiple endpoints, in which case only some endpoints get triggered. This method allows the form to be updated partially, instead of requiring the entire new state of the form via `resetTo`.
- `receiveErrors: (errors?) => void`, to support automatic binding of back-end validation errors with `setReactHookFormError` and toasting on network errors.
- `mutate: (data) => void`, to support updating parts of the form, setting new fields as the new defaults, _without touching the existing state of the form_, except the new fields.

The documentations for these APIs are written in TSDoc in `Form.tsx`. I am leaving these here for the technical story.

`mutate` may sound like a very niche feature, but it is quite useful. I developed it for a use case where one wants to update _the defaults_ of some fields and _immediately show them_, but leaving the other fields untouched. It is used when one modifies the Emails section in Account Settings. For example, when adding a new email, the `emails` field of the form needs to be updated, but the rest of the form may have edits. These changes are not saved because adding an email is a POST handled by a different endpoint. The new `emails` data must be set as the new defaults.

purfectliterature commented:
Do we know exactly what the behaviours of the email types are? I was thinking that I'd be good to list them down so users are clear about how we are using their emails.

For example, all confirmed emails can be used to log in, so I listed it out. Others may include which emails that receives notifications, etc.

Is there something else that a Primary email can do that other confirmed emails cannot?

![image](https://user-images.githubusercontent.com/51525686/197342084-3c576773-f173-4190-9ebd-c32b2024e9df.png)

purfectliterature commented:
Do you think it is better to highlight the confirmed emails instead of the unconfirmed ones? I thought since we already have the yellow `Alert` for unconfirmed emails, it has received enough emphasis. On the other hand, confirmed emails do not have a definitive status label.

![image](https://user-images.githubusercontent.com/51525686/197342325-ae56eeed-d8ca-47f9-bcc2-4d4d1701fd5d.png)

Another option is to have badges for every emails. In the below example, I changed the unconfirmed badge to outline to de-emphasise it against the primary badge. An argument against this design is that there might be too much visual clutter, but it does look consistent.

![image](https://user-images.githubusercontent.com/51525686/197342418-d0955b3e-47d5-4051-8ae7-a59f08cf63ab.png)

ekowidianto reviewed (approved):
Good work!
Some nit comments below and:
- After uploading and saving a new profile picture, the component reverts to the old picture. Only after refreshing can we see the updated one
- Disable pasting in `Confirm New Password` field.
- Re-use `ImageCropDialog` in course settings

purfectliterature commented:
The password fields work fine with Google Password Manager. Not sure why it did not pop up in the demo earlier today.

![Screenshot 2022-10-26 at 4 57 45 PM](https://user-images.githubusercontent.com/51525686/197982552-e1e8a03c-dcd9-4ba4-a225-7797bd7be701.png)

purfectliterature commented on client/app/lib/components/core/fields/TextField.tsx:
```diff
@@ -9,16 +9,30 @@ import {
 
 type TextFieldProps = ComponentProps<typeof MuiTextField> & {
   trims?: boolean;
+  onChangePasswordVisibility?: (visibility: boolean) => void;
```
Separating `PasswordTextField` was surprisingly not trivial. I imagined a future where we might have so many different descendants of `TextField` that we can no longer deal with simple if-clauses.

I tried to make `core/fields/TextField` that can have dynamic props based on the `type` (i.e., if `number`, extend `ComponentProps<typeof NumberTextField>`), but the typing was so complicated and `createElement` for some reasons keep rejecting my mapped `*TextField`s.

For now, we are stuck with `FormTextField` that has dynamic props, but internally still uses if-clauses to resolve the concrete `*TextField` elements; 50% beauty achieved(?)

>**Note** The type casting of `elementProps` in `FormTextField` is necessary because for some reasons, TypeScript keeps complaining that the `variant` prop is incompatible, despite it not being extracted/modified anywhere. I think this is one of TypeScripts edge cases; I have seen similar cases in my friends' projects.

I think the best solution is to not deal with this at all. We can continue to have loose `*TextField`s that compose `core/fields/TextField`, but instead of having `FormTextField`, we can just have a generic `FormTextField` or `FieldOf` components that works like:
```jsx
<FieldOf kind={PasswordTextField} ... /> // <-- props here is of PasswordTextField's
```
or
```jsx
// `kind` defaults to core/fields/TextField
<FormTextField kind={PasswordTextField} ... /> // <-- props here is of PasswordTextField's
```
The former is better in general because it decouples our core components from our forms, but may take some time to clear out `form/fields` and everyone that uses it. The latter is easier is faster to implement 🤞 but does not really solve the root of the problem.

Both of these solutions, however, expose the 'internality' of our core components. Our current implementation encapsulates whatever `*TextField`, which is good.

Anyway, thought it'd be nice to document this here.

purfectliterature commented:
New changes since last review:
- Disable GET `users/edit` and `user/emails/edit` routes
- Sort emails by primary and confirmed time
- Add email subsection is now collapsible
- Add email field now resets when the form resets
- Replace email abilities with badges
- Fix `Form` heads-up appearing above `Dialog`s

ekowidianto reviewed (approved):
Good work and thanks for the update!
